### ISSUE TITLE
Spanner backup integration test timeout

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp#4306]
https://source.cloud.google.com/results/invocations/a6f0c51a-80ee-4690-886a-71476ad4b87f/targets

We should probably fix this by running the "integration" tests (which use bazel) w/ the Spanner emulator if we can.

**Comment by devjgm** [comment]
Oh, wait this suggests that Spanner backups are not supported by the emulator.

https://github.com/googleapis/google-cloud-cpp/blob/080825e2387de8688fd0a7e5c5970f04d2d118c0/google/cloud/spanner/integration_tests/backup_integration_test.cc#L92-L94

Need a new plan.

**Comment by devjgm** [comment]
I'm reverting the PR that "fixed" this. Revert PR: https://github.com/googleapis/google-cloud-cpp/pull/4316

Increasing the bazel timeout did not work because the policy still had a 15 minute timeout:

**Comment by devjgm** [comment]
Before re-enabling the Spanner backup tests, we should:

1. Understand why/where all the time is going.
2. See if we can adjust the timeouts used in the policy objects to reduce the total time or at least remove the flakiness
3. Re-enable the tests only once (we think) the flakiness (i.e., timeouts) are gone or very rare.

**Comment by devjgm** [comment]
I added some timing and logging to the https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/spanner/integration_tests/backup_integration_test.cc to see which functions are taking the longest.

**Raw output:**
```
[ RUN      ] BackupTestWithCleanup.BackupTestSuite
2020-05-29T22:09:10.997560838+00:00 ### RUNNING BACKUP TESTS
2020-05-29T22:09:11.075771304+00:00 ### CreateInstance()...
2020-05-29T22:09:12.511216361+00:00 ### DONE
2020-05-29T22:09:16.768812067+00:00 ### CreateBackup()...
2020-05-29T22:09:21.514117866+00:00 ### got future
2020-05-29T22:09:22.835259606+00:00 ### cancelling...
2020-05-29T22:09:23.42027205+00:00 ### DONE
2020-05-29T22:09:23.420340957+00:00 ### CreateBackup() (again)...
2020-05-29T22:09:25.980628454+00:00 ### ListBackupOperationrs()...
2020-05-29T22:09:26.150394747+00:00 ### DONE
2020-05-29T22:09:26.150484425+00:00 ### Getting backup_future.get()...
2020-05-29T22:20:15.235613176+00:00 ### DONE
2020-05-29T22:20:15.235746058+00:00 ### GetBackup()...
2020-05-29T22:20:15.392387452+00:00 ### DONE
2020-05-29T22:20:15.392452589+00:00 ### RestoreDatabase()...
2020-05-29T22:26:56.066126258+00:00 ### DONE
2020-05-29T22:26:56.066232534+00:00 ### ListDatabaseOperations()...
2020-05-29T22:26:56.234138138+00:00 ### DONE
2020-05-29T22:26:56.234200238+00:00 ### DropDatabase()...
2020-05-29T22:26:57.111009647+00:00 ### DONE
2020-05-29T22:26:57.111094466+00:00 ### ListBackups()...
2020-05-29T22:26:57.2819705+00:00 ### DONE
2020-05-29T22:26:57.282048358+00:00 ### UpdateBackupExpireTime()...
2020-05-29T22:26:57.747238619+00:00 ### DONE
2020-05-29T22:26:57.747324185+00:00 ### DeleteBackup()...
2020-05-29T22:26:59.312201863+00:00 ### DONE
2020-05-29T22:26:59.312275858+00:00 ### DeleteInstance()...
2020-05-29T22:27:11.200418838+00:00 ### DONE
[       OK ] BackupTestWithCleanup.BackupTestSuite (1080456 ms)
```

**Summary of slow operations:**
* `CreateInstance()` takes ~2 min
* `CreateBackup()` *cancelled* takes ~ 7 min
* `CreateBackup()` *waiting for completion* takes ~ 11 min
* `RestoreDatabase()` takes ~ 7 min
* `DropDatabase()` takes ~ 1 min

**Comment by coryan** [comment]
> I added some timing and logging to the https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/spanner/integration_tests/backup_integration_test.cc to see which functions are taking the longest.

Did you try just setting `GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes` and `GOOGLE_CLOUD_CPP_SPANNER_ENABLE_TRACING=rpc`?  I think that would give you the timing you want without code changes.  And will probably break down the `CreateBackup()` step a bit more.

We should try just running a backup without canceling and see if that takes a long time. It is possible that the two operations are serializing in the backend or something.

If that is true, then using different databases (and even instances!) for `CreateBackup() + cancel` vs. `CreateBackup() + wait-for-completion` might solve the issue.

**Comment by devjgm** [comment]
> > I added some timing and logging to the https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/spanner/integration_tests/backup_integration_test.cc to see which functions are taking the longest.
> 
> Did you try just setting `GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes` and `GOOGLE_CLOUD_CPP_SPANNER_ENABLE_TRACING=rpc`? I think that would give you the timing you want without code changes. And will probably break down the `CreateBackup()` step a bit more.

Now that you suggest it, yes, I did that. It confirmed that we backoff to about a max of a 1 min delay, then keep retrying at ~1-min intervals. That's because of this code: https://github.com/googleapis/google-cloud-cpp/blob/fcc3103fb0b639e209cf396370ca5d33fb83f474/google/cloud/spanner/testing/policies.h#L38-L52

So it looks like the long backup times are not because of our retries hitting a long poll interval.

> 
> We should try just running a backup without canceling and see if that takes a long time. It is possible that the two operations are serializing in the backend or something.

Good idea. I'll try that and see what happens.

> 
> If that is true, then using different databases (and even instances!) for `CreateBackup() + cancel` vs. `CreateBackup() + wait-for-completion` might solve the issue.

**Comment by mr-salty** [comment]
Note when CMEK is merged in, I added another integration test, similar to but somewhat simpler than the original:

* It uses `PickRandomInstance` instead of creating its own (the original probably could too?)
* It does not do the Cancel
* It doesn't do anything with `ListBackupOperations`, it just waits for completion.
* It *does* drop both the original and restored databases. The existing test doesn't do the former, which probably doesn't matter since it deletes the instance, but it should if it uses `PickRandomInstance`.

The test took 585s (=~10min) to run against spanner staging; the feature is not yet available in prod but brad thought prod was generally faster.  I don't have individual operation timings though.

**Comment by coryan** [comment]
A repeat of this:

https://pantheon.corp.google.com/cloud-build/builds;region=us-east1/593a64c0-e078-4210-aa94-3d9eac712294;step=3?project=cloud-cpp-testing-resources

**Comment by devbww** [comment]
The set of CreateBackup() calls not completing within the one hour we've allotted is probably not actionable by us. As far as I understand things, these operations are batch-scheduled in the backend, so service time is unpredictable.  We could be blocked behind someone slow.

Given that these are the "daily slow" tests, I'm not sure that is a problem, but if the timeout frequency is too high we could increase the timeout, or refactor/filter the test so that each CreateBackup() gets its own googletest-level timeout.

It certainly might be nice to know we were actually in a CreateBackup() poll loop when the timeout happened, but I'm unaware of any hook we could use to dump the logging buffer when that happens.

**Comment by coryan** [comment]
> It certainly might be nice to know we were actually in a CreateBackup() poll loop when the timeout happened, but I'm unaware of any hook we could use to dump the logging buffer when that happens.

Maybe a separate thread that dumps the logs at `timeout - 60 seconds` unless prevented by the test suite tear down?

> Given that these are the "daily slow" tests, I'm not sure that is a problem, but if the timeout frequency is too high we could increase the timeout, or refactor/filter the test so that each CreateBackup() gets its own googletest-level timeout.

nit: it my book that makes the problem "actionable" 🤷

**Comment by devbww** [comment]
Allow me to correct a misconception I may have created in my previous message.  There is no such thing as a _"googletest-level timeout"_.  Test timeouts are only in the `bazel test`/`ctest` domain.  As far as I know, there is no visibility into that timeout from the code under test, and therefore no way to determine something like _"timeout - 60 seconds"_.

Let me also take this opportunity to correct some timings from @devjgm's _"**Summary of slow operations**"_ message, where there appears to have been some confusion between minutes and seconds.  The timings look like they are actually:

**Summary of slow operations:**

- CreateInstance() takes ~2 sec
- CreateBackup() cancelled takes ~ 7 sec
- CreateBackup() waiting for completion takes ~ 11 min
- RestoreDatabase() takes ~ 7 min
- DropDatabase() takes ~ 1 sec

That is, only (uncancelled) `CreateBackup()` and `RestoreDatabase()` are "slow".

**Comment by devjgm** [comment]
Another backup timeout during the daily integration test:
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/7cee8077-529f-4699-9466-74a94a08f040;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/95d00e02df1e626273352141411f6d5664033c7d/fedora-34-integration-daily/log-7cee8077-529f-4699-9466-74a94a08f040.txt

**Comment by coryan** [comment]
https://pantheon.corp.google.com/cloud-build/builds;region=us-east1/c07cb6ab-1ad9-4a5f-bd40-1b327619e2e1?project=cloud-cpp-testing-resources

**Comment by coryan** [comment]
https://pantheon.corp.google.com/cloud-build/builds;region=us-east1/cdc748b8-8c18-4a68-817e-7a619df54ac3;step=3?project=cloud-cpp-testing-resources

**Comment by devbww** [comment]
https://console.cloud.google.com/cloud-build/builds;region=us-east1/88ae524b-eeac-43d5-bc7d-b8898f198f29?project=cloud-cpp-testing-resources

In this run, before `BackupTest.BackupTestWithCMEK` timed out at the 1h limit, we had `BackupTest.BackupTest` taking about 32m and `BackupTest.CreateBackupWithVersionTime` taking about 17m.

**Comment by devbww** [comment]
For the record, in the two days since #7195 landed we've seen ...
```
2021-08-20:
INFO: Elapsed time: 4143.082s, Critical Path: 4142.48s
//google/cloud/spanner/benchmarks:benchmarks_config_test                 PASSED in 0.2s
//google/cloud/spanner/benchmarks:multiple_rows_cpu_benchmark            PASSED in 146.2s
//google/cloud/spanner/benchmarks:single_row_throughput_benchmark        PASSED in 33.5s 
//google/cloud/spanner/integration_tests:backup_integration_test         PASSED in 4142.4s
//google/cloud/spanner/integration_tests:client_integration_test         PASSED in 19.7s 
//google/cloud/spanner/integration_tests:client_stress_test              PASSED in 33.6s 
//google/cloud/spanner/integration_tests:data_types_integration_test     PASSED in 21.3s 
//google/cloud/spanner/integration_tests:database_admin_integration_test PASSED in 142.5s
//google/cloud/spanner/integration_tests:instance_admin_integration_test PASSED in 37.1s 
//google/cloud/spanner/integration_tests:rpc_failure_threshold_integration_test PASSED in 29.3s 
//google/cloud/spanner/integration_tests:session_pool_integration_test   PASSED in 19.5s 
//google/cloud/spanner/samples:samples                                   PASSED in 2662.3s
```
and ...
```
2021-08-21:
INFO: Elapsed time: 3496.377s, Critical Path: 3495.76s
//google/cloud/spanner/benchmarks:benchmarks_config_test                 PASSED in 0.2s
//google/cloud/spanner/benchmarks:multiple_rows_cpu_benchmark            PASSED in 147.3s
//google/cloud/spanner/benchmarks:single_row_throughput_benchmark        PASSED in 27.8s 
//google/cloud/spanner/integration_tests:backup_integration_test         PASSED in 3495.7s
//google/cloud/spanner/integration_tests:client_integration_test         PASSED in 32.3s 
//google/cloud/spanner/integration_tests:client_stress_test              PASSED in 23.8s 
//google/cloud/spanner/integration_tests:data_types_integration_test     PASSED in 20.0s 
//google/cloud/spanner/integration_tests:database_admin_integration_test PASSED in 164.2s
//google/cloud/spanner/integration_tests:instance_admin_integration_test PASSED in 37.1s 
//google/cloud/spanner/integration_tests:rpc_failure_threshold_integration_test PASSED in 42.8s 
//google/cloud/spanner/integration_tests:session_pool_integration_test   PASSED in 26.7s 
//google/cloud/spanner/samples:samples                                   PASSED in 2440.7s
```
so the first of those would have timed out with the old 3600s limit.

**Comment by coryan** [comment]
https://pantheon.corp.google.com/cloud-build/builds;region=us-east1/94b9fc61-64c8-4352-adac-785d07fc1bf7;step=3?project=cloud-cpp-testing-resources

**Comment by coryan** [comment]
https://pantheon.corp.google.com/cloud-build/builds;region=us-east1/3962c5fb-bb4f-42f5-a792-d3c997875b36;step=3?project=cloud-cpp-testing-resources

**Comment by coryan** [comment]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/e65a5bc5-6281-4c37-9118-52e59f193092;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/f5e33f31fb2fcb39e726caffeeb1c4d593b8d13b/fedora-34-integration-daily/log-e65a5bc5-6281-4c37-9118-52e59f193092.txt

**Comment by dbolduc** [comment]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/cf8747da-4974-4be9-b788-95c2bac7897d;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/6682bb40ad56257cffb687f3052a80e0f2bfe3ea/fedora-34-integration-daily/log-cf8747da-4974-4be9-b788-95c2bac7897d.txt

**Comment by devbww** [comment]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/4907495a-26cb-4ba7-ac27-a842f8cae8ae;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/38c626034c4a3e5a07ff1b47d05fdf2ec8c6f459/fedora-34-integration-daily/log-4907495a-26cb-4ba7-ac27-a842f8cae8ae.txt

**Comment by devbww** [comment]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/b010df75-dacc-4d18-b403-6cd6a6858542;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/b9e43534d8349e608f32486a1b922275eb0f1819/fedora-34-integration-daily/log-b010df75-dacc-4d18-b403-6cd6a6858542.txt

**Comment by dbolduc** [comment]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/9bfa4b78-a955-4cb6-b3df-0018c8930071;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/555dbccd405d1cdeebcfff8ae7a06f067228e939/fedora-34-integration-daily/log-9bfa4b78-a955-4cb6-b3df-0018c8930071.txt

**Comment by devbww** [comment]
As of today we should only be running the following tests against the listed instances:
```
//google/cloud/spanner/admin/integration_tests:backup_integration_test            test-instance-12
//google/cloud/spanner/admin/integration_tests:backup_extra_integration_test      test-instance-13
//google/cloud/spanner/integration_tests:backup_integration_test                  test-instance-14
//google/cloud/spanner/integration_tests:backup_extra_integration_test            test-instance-15
```
All of those instances are in `us-east1`, and so we should be avoiding the current contention in `us-central1`.

**Comment by devbww** [comment]
That last instance of Dec 14, 2021, has been recategorized as https://github.com/googleapis/google-cloud-cpp/issues/7741.

**Comment by devbww** [comment]
We have mitigations in effect for tolerating polling timeouts in long-running `CreateBackup()` operations, at least until we get some guidance from the backend folks about expected service times.  But also closing this in favor of #8616.
